00-programming-fundamentals/debugging-with-vs-code.md

```diff
@@ -0,0 +1,177 @@
+
+For every two developers, there are three opinions about the best environment for making code. So how do we pick? It all depends on how much help (some would say interference) a programmer wants moment to moment.
+
+On the most basic level, a Text Editor will do just that: let you edit the words and symbols on a page unfettered by suggestion text or drop downs. An Integrated Development Environment (IDE for short) on the other hand has opinions about how you should be doing your work: they may try and give you the name of the function you were about to type, help manage a source control tool, or (as we will do) help you debug your code.
+
+__Note: If anyone ever tells you that "real developers only use [their personal favorite tool]", laugh in their face and walk away from that conversation.__
+
+## Setting Up
+The relationship between VS Code and Ruby is still in its infancy, so many of the tools that we regularly use in other languages (such as C++, Javascript or Ook) aren't immidiately accessible. Our process will be:
+
+1. Set up the file we want to debug
+1. Make sure we have the proper gems on our machine
+1. Configure our `launch.json` in our `.vscode` folder
+1. Set our breakpoint
+1. Debug our code
+
+Copy and paste the following code into a new `.rb` file:
+
+```ruby
+#my_efficient_and_clever_program.rb
+
+#definitely always sums the two numbers you give it.
+def cool_sum_func(a, b)
+    puts "a = #{a}"
+    puts "b = #{b}"
+
+    #completely necessary if-else
+    if (a%2 == 0)
+        return add_them(a, b)
+    else
+        b += 2
+        return subtract_them(a,b)
+    end
+end
+
+#adds two numbers
+def add_them(a, b)
+    return a + b
+end
+
+#subtracts b from a
+def subtract_them(a, b)
+    return a - b
+end
+
+
+left_num = 5
+right_num = 3
+
+sum = cool_sum_func(left_num, right_num)
+
+puts sum
+```
+
+This is some very dopey code, but it will show us a lot about how the debugger gives us information.
+
+## Get Relevant Gems and Extensions
+If you have not already run the following:
+
+```bash
+$:gem install ruby-debug-ide
+$:gem install debase
+```
+
+Please run them now.
+
+## Configuring the Debugger
+
+VS Code has a very bare-bones debugger that comes with it, but it doesn't know how to talk to every language out of the box. When we set up projects in VS Code, we create a series of configurations that VS Code stores in a file called `launch.json` in a folder called `.vscode`.
+
+Open the debugger by using `⇧⌘D`, and look near the top left of the viewport to find a green triangle button that likely reads "No Configurations". Click on the text and select "Add Configuration" from the drop down menu. It doesn't matter what you pick after this, as we will be filling in the file for ourselves from this point forward. When you have finished selecting a pre-generated option, a new pane should open called `launch.json`. Replace the file with the following:
+
+```json
```

Comment:
Seems like GitHub particularly hates formatting this-- I haven't tried, but I might try using `javascript` syntax instead for this particular block?